src/ZoneServer/World/Storage/Storage.cs

```diff
@@ -0,0 +1,691 @@
+			if (storageResult != StorageResult.Success)
+				return storageResult;
+			Send.ZC_ITEM_ADD(interactingChar, itemToStore, addedPosition, itemToStore.Amount, InventoryAddType.New, invType);
+
+			return storageResult;
+		}
+
+		/// <summary>
+		/// Adds an item to storage and removes from given character.
+		/// Updates client.
+		/// </summary>
+		/// <param name="character">Character that is performing interaction</param>
+		/// <param name="objectId">ObjectId of item to store</param>
+		/// <param name="amount">Amount of item to store</param>
+		/// <param name="invType">Storage inventory type</param>
+		/// <returns></returns>
+		protected StorageResult StoreItem(Character character, long objectId, int amount, InventoryType invType)
+		{
+			var item = character.Inventory.GetItem(objectId);
+			if (item == null)
+				return StorageResult.ItemNotFound;
+
+			Item itemToStore = new Item(item);
+			itemToStore.Amount = Math.Min(itemToStore.Amount, amount);
+
+			// Checks amount that can be stored
+			var availableAmount = this.GetItemAvailableAmount(itemToStore);
+			if (availableAmount <= 0)
+				return StorageResult.StorageFull;
+			itemToStore.Amount = Math.Min(itemToStore.Amount, availableAmount);
+
+			// Remove the amount from character
+			var inventoryResult = character.Inventory.Remove(item, itemToStore.Amount, InventoryItemRemoveMsg.Given);
+			if (inventoryResult != InventoryResult.Success)
+				return StorageResult.InvalidOperation;
+
+			// Add the amount to storage
+			var addResult = this.AddItemAvailableAmount(character, itemToStore, invType);
+			if (addResult != StorageResult.Success)
+				return addResult;
+
+			return StorageResult.Success;
+		}
+
+		/// <summary>
+		/// Adds a list of items to storage and removes from given character.
+		/// The amount added is always the maximum item amount.
+		/// Updates client.
+		/// </summary>
+		/// <param name="character">Character that is performing interaction</param>
+		/// <param name="objectId">ObjectId of item to store</param>
+		/// <param name="invType">Storage inventory type</param>
+		/// <returns></returns>
+		protected StorageResult StoreItems(Character character, List<long> objectIds, InventoryType invType)
+		{
+			foreach (var objectId in objectIds)
+			{
+				var item = character.Inventory.GetItem(objectId);
+				if (item == null)
+					return StorageResult.ItemNotFound;
+
+				var result = StoreItem(character, objectId, item.Amount, invType);
+				if (result != StorageResult.Success)
+					return result;
+			}
+
+			return StorageResult.Success;
+		}
+
+		/// <summary>
+		/// Removes an item from storage and gives to a character.
+		/// Updates client.
+		/// </summary>
+		/// <param name="character"></param>
+		/// <param name="objectId"></param>
+		/// <param name="amount"></param>
+		/// <param name="invType"></param>
+		/// <returns></returns>
+		protected StorageResult RetrieveItem(Character character, long objectId, int amount, InventoryType invType)
```

Comment:
Similar to storing this feels somewhat restrictive, not allowing us to get an item without adding it to a character's inventory. I realize this is (probably) not a function we'll need during normal gameplay, but I could think of a couple reasons why someone might want to do that, be it for administrative purposes, support, or even events.

Wouldn't it be relatively simple to make the storage more of a general purpose item list that the child classes than expand to do what they need? That would arguably make for a much more flexible system, and it would also remove certain patterns, like passing the character to the storage for getting items or sending updates.

Reply:
I'm not sure if this was changed or what, but after looking again at my code I realized this is not the case.

In Storage.cs class I am providing 'RetrieveItem()' as an easy way to retrieve the item **for a character** since it will also send packets to the character received in parameter.

However, this is not the only method that allows you to get items from storage. You can also call GetItem() method which is public and allows for retrieval of items in storage without need of character parameter. Same for Add() and AddAtPosition() methods, which are all public.